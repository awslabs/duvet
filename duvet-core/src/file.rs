// Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
// SPDX-License-Identifier: Apache-2.0

use crate::{contents::Contents, diagnostic::IntoDiagnostic, path::Path, Result};
use core::{
    fmt,
    ops::{Deref, Range},
};
use miette::{SourceCode, WrapErr};
use std::sync::Arc;

#[derive(Clone, Debug, PartialEq, Eq, PartialOrd, Ord, Hash)]
pub struct BinaryFile {
    pub(crate) path: Path,
    pub(crate) contents: Contents,
}

impl BinaryFile {
    pub fn new<P, C>(path: P, contents: C) -> Self
    where
        P: Into<Path>,
        C: Into<Contents>,
    {
        let path = path.into();
        let contents = contents.into();
        Self { path, contents }
    }

    pub fn path(&self) -> &Path {
        &self.path
    }

    pub fn hash(&self) -> &[u8; 32] {
        self.contents.hash()
    }
}

impl Deref for BinaryFile {
    type Target = [u8];

    fn deref(&self) -> &Self::Target {
        &self.contents
    }
}

#[derive(Clone, Debug, PartialEq, Eq, PartialOrd, Ord, Hash)]
pub struct SourceFile {
    pub(crate) path: Path,
    pub(crate) contents: Contents,
}

impl SourceFile {
    pub fn new<P, C>(path: P, contents: C) -> Result<Self>
    where
        P: Into<Path>,
        C: Into<Contents>,
    {
        let path = path.into();
        let contents = contents.into();
        let _ = core::str::from_utf8(&contents)
            .into_diagnostic()
            .wrap_err_with(|| path.clone())?;
        Ok(Self { path, contents })
    }

    pub fn path(&self) -> &Path {
        &self.path
    }

    pub fn hash(&self) -> &[u8; 32] {
        self.contents.hash()
    }

    pub async fn as_toml<T>(&self) -> crate::Result<Arc<T>>
    where
        T: 'static + Send + Sync + serde::de::DeserializeOwned,
    {
        let path = self.path.clone();
        let contents = self.contents.clone();
        // TODO can we get better errors by mapping string ranges?
        crate::Cache::current()
            .get_or_init(*self.hash(), move || {
                crate::Query::from(
                    toml_edit::de::from_slice(contents.data())
                        .map(Arc::new)
                        .into_diagnostic()
                        .wrap_err(path)
                        .map_err(|err| err.into()),
                )
            })
            .await
    }

    pub async fn as_json<T>(&self) -> crate::Result<Arc<T>>
    where
        T: 'static + Send + Sync + serde::de::DeserializeOwned,
    {
        let path = self.path.clone();
        let contents = self.contents.clone();
        // TODO can we get better errors by mapping string ranges?
        crate::Cache::current()
            .get_or_init(*self.hash(), move || {
                crate::Query::from(
                    serde_json::from_slice(contents.data())
                        .map(Arc::new)
                        .into_diagnostic()
                        .wrap_err(path)
                        .map_err(|err| err.into()),
                )
            })
            .await
    }

    pub fn substr(&self, v: &str) -> Option<Slice<SourceFile>> {
        unsafe {
            let beginning = self.as_bytes().as_ptr();
            let end = beginning.add(self.as_bytes().len());

            if !(beginning..=end).contains(&v.as_ptr()) {
                return None;
            }

            Some(self.substr_unchecked(v))
        }
    }

    /// # Safety
    ///
    /// Callers should ensure that the `v` is a slice of `self`
    pub unsafe fn substr_unchecked(&self, v: &str) -> Slice<SourceFile> {
        let start = v.as_bytes().as_ptr() as usize - self.as_bytes().as_ptr() as usize;
        Slice {
            file: self.clone(),
            start,
            end: start + v.len(),
        }
    }
}

impl Deref for SourceFile {
    type Target = str;

    fn deref(&self) -> &Self::Target {
        unsafe {
            // Safety: this validity was checked at creation on SourceFile
            core::str::from_utf8_unchecked(&self.contents)
        }
    }
}

impl AsRef<str> for SourceFile {
    fn as_ref(&self) -> &str {
        self
    }
}

impl SourceCode for SourceFile {
    fn read_span<'a>(
        &'a self,
        span: &miette::SourceSpan,
        context_lines_before: usize,
        context_lines_after: usize,
    ) -> Result<Box<dyn miette::SpanContents<'a> + 'a>, miette::MietteError> {
        use miette::MietteSpanContents;

        let contents = (**self).read_span(span, context_lines_before, context_lines_after)?;

        let path = std::env::current_dir()
            .ok()
            .and_then(|cwd| self.path.strip_prefix(cwd).ok())
            .unwrap_or(&self.path)
            .display()
            .to_string();

        Ok(Box::new(MietteSpanContents::new_named(
            path,
            contents.data(),
            *contents.span(),
            contents.line(),
            contents.column(),
            contents.line_count(),
        )))
    }
}

#[derive(Clone, PartialEq, PartialOrd, Hash, Eq, Ord)]
pub struct Slice<File = SourceFile> {
    file: File,
    start: usize,
    end: usize,
}

impl<F: File> Slice<F> {
    pub fn path(&self) -> &Path {
        self.file.path()
    }

    pub fn file(&self) -> &F {
        &self.file
    }

    pub fn range(&self) -> Range<usize> {
        self.start..self.end
    }
}

impl Slice<SourceFile> {
    pub fn error<E>(&self, e: E, label: impl AsRef<str>) -> crate::diagnostic::Error
    where
        E: Into<crate::diagnostic::Error>,
    {
        e.into().with_source_slice(self.clone(), label)
    }
}

impl fmt::Debug for Slice<BinaryFile> {
    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
        self[..].fmt(f)
    }
}

impl fmt::Debug for Slice<SourceFile> {
    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
        self[..].fmt(f)
    }
}

impl fmt::Display for Slice<SourceFile> {
    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
        self[..].fmt(f)
    }
}

impl Deref for Slice<BinaryFile> {
    type Target = [u8];

    fn deref(&self) -> &Self::Target {
        unsafe {
            // Safety: range validatity was checked at slice creation time
            self.file.get_unchecked(self.start..self.end)
        }
    }
}

impl AsRef<[u8]> for Slice<BinaryFile> {
    #[inline]
    fn as_ref(&self) -> &[u8] {
        self
    }
}

impl Deref for Slice<SourceFile> {
    type Target = str;

    fn deref(&self) -> &Self::Target {
        unsafe {
            // Safety: range validatity was checked at slice creation time
            self.file.get_unchecked(self.start..self.end)
        }
    }
}

impl AsRef<str> for Slice<SourceFile> {
    #[inline]
    fn as_ref(&self) -> &str {
        self
    }
}

impl PartialEq<[u8]> for Slice<BinaryFile> {
    fn eq(&self, other: &[u8]) -> bool {
        (**self).eq(other)
    }
}

impl PartialEq<[u8]> for Slice<SourceFile> {
    fn eq(&self, other: &[u8]) -> bool {
        (**self).as_bytes().eq(other)
    }
}

impl PartialEq<str> for Slice<SourceFile> {
    fn eq(&self, other: &str) -> bool {
        (**self).eq(other)
    }
}

pub trait File {
    fn path(&self) -> &Path;
}

impl File for SourceFile {
    fn path(&self) -> &Path {
        &self.path
    }
}

impl File for BinaryFile {
    fn path(&self) -> &Path {
        &self.path
    }
}
