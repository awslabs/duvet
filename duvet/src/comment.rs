// Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
// SPDX-License-Identifier: Apache-2.0

use crate::{
    annotation::{AnnotationSet, AnnotationType},
    Error, Result,
};
use duvet_core::{error, file::SourceFile};
use std::sync::Arc;

pub mod parser;
pub mod tokenizer;

#[cfg(test)]
mod tests;

pub fn extract(
    file: &SourceFile,
    pattern: &Pattern,
    default_type: AnnotationType,
) -> (AnnotationSet, Vec<Error>) {
    let tokens = tokenizer::tokens(file, pattern);
    let mut parser = parser::parse(tokens, default_type);

    let annotations = (&mut parser).collect();
    let errors = parser.errors();

    (annotations, errors)
}

#[derive(Clone, Debug, PartialEq, PartialOrd, Eq, Ord, Hash)]
pub struct Pattern {
    pub meta: Arc<str>,
    pub content: Arc<str>,
}

impl Default for Pattern {
    fn default() -> Self {
        Self {
            meta: "//=".into(),
            content: "//#".into(),
        }
    }
}

impl Pattern {
    pub fn from_arg(arg: &str) -> Result<Self> {
        let mut parts = arg.split(',').filter(|p| !p.is_empty());
        let meta = parts.next().expect("should have at least one pattern");
        if meta.is_empty() {
            return Err(error!("compliance pattern cannot be empty"));
        }

        let content = parts.next().unwrap();

        let meta = meta.into();
        let content = content.into();

        Ok(Self { meta, content })
    }
}
