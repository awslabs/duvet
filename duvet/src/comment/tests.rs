// Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
// SPDX-License-Identifier: Apache-2.0

use super::*;

fn parse(pattern: &str, value: &str) -> (AnnotationSet, Vec<String>) {
    let file = SourceFile::new("file.rs", value).unwrap();
    let pattern = Pattern::from_arg(pattern).unwrap();
    let (annotations, errors) = extract(&file, &pattern, Default::default());
    let errors = errors.into_iter().map(|error| error.to_string()).collect();
    (annotations, errors)
}

macro_rules! snapshot {
    ($name:ident, $value:expr) => {
        snapshot!($name, "//=,//#", $value);
    };
    ($name:ident, $pattern:expr, $value:expr) => {
        #[test]
        fn $name() {
            let mut settings = insta::Settings::clone_current();
            // ignore CWD
            settings.add_filter(
                &dbg!(duvet_core::env::current_dir()
                    .unwrap()
                    .as_ref()
                    .display()
                    .to_string()
                    .replace('/', "\\/")),
                "[CWD]",
            );
            let _bound = settings.bind_to_scope();
            insta::assert_debug_snapshot!(stringify!($name), parse($pattern, $value));
        }
    };
}

snapshot!(
    content_without_meta,
    r#"
    //# This is some content without meta
    "#
);

snapshot!(
    meta_without_content,
    r#"
    //= type=todo
    "#
);

snapshot!(
    type_citation,
    r#"
    //= https://example.com/spec.txt
    //# Here is my citation
    "#
);

snapshot!(
    type_test,
    r#"
    //= https://example.com/spec.txt
    //= type=test
    //# Here is my citation
    "#
);

snapshot!(
    type_todo,
    r#"
    //= https://example.com/spec.txt
    //= type=todo
    //= feature=cool-things
    //= tracking-issue=123
    //# Here is my citation
    "#
);

snapshot!(
    type_exception,
    r#"
    //= https://example.com/spec.txt
    //= type=exception
    //= reason=This isn't possible currently
    //# Here is my citation
    "#
);

snapshot!(
    missing_new_line,
    r#"
    //= https://example.com/spec.txt
    //# Here is my citation"#
);
