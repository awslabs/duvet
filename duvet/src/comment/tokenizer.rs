// Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
// SPDX-License-Identifier: Apache-2.0

use crate::comment::Pattern;
use duvet_core::file::{Slice, SourceFile};

#[derive(Clone, Debug, PartialEq, Eq)]
pub enum Token {
    Meta {
        key: Slice<SourceFile>,
        value: Slice<SourceFile>,
        line: usize,
    },
    UnnamedMeta {
        value: Slice<SourceFile>,
        line: usize,
    },
    Content {
        value: Slice<SourceFile>,
        line: usize,
    },
}

impl Token {
    pub fn line_no(&self) -> usize {
        match self {
            Token::Meta { line, .. } => *line,
            Token::UnnamedMeta { line, .. } => *line,
            Token::Content { line, .. } => *line,
        }
    }
}

pub fn tokens<'a>(file: &'a SourceFile, style: &'a Pattern) -> Tokenizer<'a> {
    Tokenizer::new(&style.meta, &style.content, file)
}

pub struct Tokenizer<'a> {
    meta_prefix: &'a str,
    content_prefix: &'a str,
    file: &'a SourceFile,
    lines: core::iter::Enumerate<core::str::Lines<'a>>,
}

impl<'a> Tokenizer<'a> {
    fn new(meta_prefix: &'a str, content_prefix: &'a str, file: &'a SourceFile) -> Self {
        Self {
            meta_prefix,
            content_prefix,
            file,
            lines: file.lines().enumerate(),
        }
    }

    fn on_line(&mut self, line: &str, line_no: usize) -> Option<Token> {
        let content = line.trim_start();
        if content.is_empty() {
            return None;
        }

        if let Some(content) = content.strip_prefix(self.meta_prefix) {
            let content = content.trim_start();
            return self.on_meta(content, line_no);
        }

        if let Some(content) = content.strip_prefix(self.content_prefix) {
            let content = content.trim_start();
            return self.on_content(content, line_no);
        }

        None
    }

    fn on_content(&mut self, content: &str, line_no: usize) -> Option<Token> {
        let value = self.file.substr(content).unwrap();
        Some(Token::Content {
            value,
            line: line_no,
        })
    }

    fn on_meta(&mut self, meta: &str, line_no: usize) -> Option<Token> {
        let mut parts = meta.trim_start().splitn(2, '=');

        let key = parts.next().unwrap();
        let key = key.trim_end();
        let key = self.file.substr(key).unwrap();

        if let Some(value) = parts.next() {
            let value = value.trim();
            let value = self.file.substr(value).unwrap();
            Some(Token::Meta {
                key,
                value,
                line: line_no,
            })
        } else {
            Some(Token::UnnamedMeta {
                value: key,
                line: line_no,
            })
        }
    }
}

impl Iterator for Tokenizer<'_> {
    type Item = Token;

    fn next(&mut self) -> Option<Self::Item> {
        loop {
            let (line_no, line) = self.lines.next()?;
            if let Some(token) = self.on_line(line, line_no) {
                return Some(token);
            }
        }
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    macro_rules! snapshot_test {
        ($name:ident, $input:expr) => {
            snapshot_test!(
                $name,
                $input,
                Pattern {
                    meta: "//@=".into(),
                    content: "//@#".into(),
                }
            );
        };
        ($name:ident, $input:expr, $config:expr) => {
            #[tokio::test]
            async fn $name() {
                let source = SourceFile::new(file!(), $input).unwrap();
                let config = $config;
                let tokens: Vec<_> = tokens(&source, &config).collect();
                insta::assert_debug_snapshot!(stringify!($name), tokens);
            }
        };
    }

    snapshot_test!(empty, "");
    snapshot_test!(
        basic,
        r#"
        //@= thing goes here
        //@= meta=foo
        //@= meta2 = bar
        //@# content goes
        //@# here
        "#
    );
    snapshot_test!(
        only_unnamed,
        r#"
        //@= this is meta
        //@= this is other meta
        "#
    );
    snapshot_test!(
        duplicate_meta,
        r#"
        //@= meta=1
        //@= meta=2
        "#
    );
    snapshot_test!(
        configured,
        r#"
        /*
         *= meta=goes here
         *# content goes here
         */
        "#,
        Pattern {
            meta: "*=".into(),
            content: "*#".into(),
        }
    );
}
