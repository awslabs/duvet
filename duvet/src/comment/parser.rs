// Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
// SPDX-License-Identifier: Apache-2.0

use super::tokenizer::Token;
use crate::{
    annotation::{Annotation, AnnotationLevel, AnnotationType},
    specification::Format,
    Error,
};
use duvet_core::{
    ensure, error,
    file::{Slice, SourceFile},
    Result,
};

pub fn parse<T: IntoIterator<Item = Token>>(
    tokens: T,
    default_type: AnnotationType,
) -> Parser<T::IntoIter> {
    Parser {
        prev_line: 0,
        meta: Default::default(),
        contents: Default::default(),
        errors: Default::default(),
        tokens: tokens.into_iter(),
        default_type,
    }
}

#[derive(Debug, Default)]
struct Meta {
    first_meta: Option<Slice>,
    target: Option<Slice>,
    anno: Option<(AnnotationType, Slice)>,
    reason: Option<Slice>,
    line: usize,
    feature: Option<Slice>,
    tracking_issue: Option<Slice>,
    level: Option<(AnnotationLevel, Slice)>,
    format: Option<(Format, Slice)>,
}

impl Meta {
    fn set(&mut self, key: Option<Slice>, value: Slice) -> Result {
        let source_value = value.clone();

        let prev = match key.as_deref() {
            Some("source") => core::mem::replace(&mut self.target, Some(value)),
            Some("level") => {
                let level = value.parse()?;
                core::mem::replace(&mut self.level, Some((level, value))).map(|v| v.1)
            }
            Some("format") => {
                let format = value.parse()?;
                core::mem::replace(&mut self.format, Some((format, value))).map(|v| v.1)
            }
            Some("type") => {
                let ty = value.parse()?;
                core::mem::replace(&mut self.anno, Some((ty, value))).map(|v| v.1)
            }
            Some("reason") => core::mem::replace(&mut self.reason, Some(value)),
            Some("feature") => core::mem::replace(&mut self.feature, Some(value)),
            Some("tracking-issue") => core::mem::replace(&mut self.tracking_issue, Some(value)),
            Some(_) => {
                return Err(key
                    .unwrap()
                    .error(error!("invalid metadata field"), "defined here"));
            }
            None => core::mem::replace(&mut self.target, Some(value)),
        };

        if let Some(prev) = prev {
            let key = key.as_deref().unwrap_or("source");
            let err = error!("{key:?} already specified")
                .with_source_slice(source_value, "redefined here")
                .with_source_slice(prev, "already defined here");
            return Err(err);
        }

        Ok(())
    }

    fn build(self, contents: Vec<Slice>, default_type: AnnotationType) -> Result<Annotation> {
        let first_meta = self.first_meta.unwrap();
        let source = first_meta.file().clone();

        let Some(target) = self.target else {
            return Err(first_meta.error(
                error!("comment is missing source specification"),
                "defined here",
            ));
        };
        let original_target = target.clone();
        let target = target.to_string();

        let anno = self.anno.map_or(default_type, |v| v.0);

        for (allowed, field) in [
            (AnnotationType::Exception, self.reason.as_ref()),
            (AnnotationType::Todo, self.tracking_issue.as_ref()),
            (AnnotationType::Todo, self.feature.as_ref()),
        ] {
            if anno != allowed {
                if let Some(value) = field {
                    return Err(value.error(error!("invalid key for type {anno}"), "defined here"));
                }
            }
        }

        let mut original_text = first_meta.range();
        let mut original_quote = original_text.clone();

        let mut quote = String::new();
        for (idx, part) in contents.iter().enumerate() {
            if idx == 0 {
                original_quote = part.range();
            } else {
                quote.push(' ');
            }

            original_text.start = original_text.start.min(part.range().start);
            original_text.end = original_text.end.max(part.range().end);
            original_quote.end = original_quote.end.max(part.range().end);

            quote.push_str(part.trim());
        }

        let original_text = source.substr_range(original_text).unwrap();
        let original_quote = source.substr_range(original_quote).unwrap();

        let annotation = Annotation {
            source: source.path().clone(),
            anno_line: self.line,
            anno,
            original_text,
            original_target,
            original_quote,
            target,
            quote,
            manifest_dir: duvet_core::env::current_dir()?,
            comment: self.reason.map(|v| v.to_string()).unwrap_or_default(),
            level: self.level.map_or(AnnotationLevel::Auto, |v| v.0),
            format: self.format.map_or(Format::Auto, |v| v.0),
            tracking_issue: self
                .tracking_issue
                .map(|v| v.to_string())
                .unwrap_or_default(),
            feature: self.feature.map(|v| v.to_string()).unwrap_or_default(),
            tags: Default::default(),
        };

        Ok(annotation)
    }
}

pub struct Parser<T> {
    prev_line: usize,
    default_type: AnnotationType,
    meta: Meta,
    contents: Vec<Slice>,
    errors: Vec<Error>,
    tokens: T,
}

impl<T: Iterator<Item = Token>> Parser<T> {
    pub fn errors(self) -> Vec<Error> {
        self.errors
    }

    fn on_token(&mut self, token: Token) -> Option<Annotation> {
        let line_no = token.line_no();
        // if the line number isn't the next expected one then flush
        let prev = self.flush_if(line_no > self.prev_line + 1);
        self.prev_line = line_no;

        match token {
            Token::Meta {
                key,
                value,
                line: _,
            } => self.push_meta(Some(key.clone()), value.clone()),
            Token::UnnamedMeta { value, line: _ } => self.push_meta(None, value.clone()),
            Token::Content { value, line: _ } => {
                self.push_contents(value.clone());
                None
            }
        }
        .or(prev)
    }

    fn push_meta(
        &mut self,
        key: Option<Slice<SourceFile>>,
        value: Slice<SourceFile>,
    ) -> Option<Annotation> {
        let prev = self.flush_if(!self.contents.is_empty());

        if self.meta.first_meta.is_none() {
            self.meta.first_meta = Some(value.clone());
            self.meta.line = (self.prev_line + 1) as _;
        }

        if let Err(err) = self.meta.set(key, value) {
            self.errors.push(err);
        }

        prev
    }

    fn push_contents(&mut self, value: Slice) {
        let file = value.file();
        let value = value.trim();
        if !value.is_empty() {
            self.contents.push(file.substr(value).unwrap());
        }
    }

    fn flush_if(&mut self, cond: bool) -> Option<Annotation> {
        if cond {
            self.flush()
        } else {
            None
        }
    }

    fn flush(&mut self) -> Option<Annotation> {
        let meta = core::mem::take(&mut self.meta);
        let contents = core::mem::take(&mut self.contents);

        ensure!(meta.first_meta.is_some(), None);

        match meta.build(contents, self.default_type) {
            Ok(anno) => Some(anno),
            Err(err) => {
                self.errors.push(err);
                None
            }
        }
    }
}

impl<T: Iterator<Item = Token>> Iterator for Parser<T> {
    type Item = Annotation;

    fn next(&mut self) -> Option<Self::Item> {
        loop {
            let Some(token) = self.tokens.next() else {
                return self.flush();
            };
            if let Some(annotation) = self.on_token(token) {
                return Some(annotation);
            }
        }
    }
}
