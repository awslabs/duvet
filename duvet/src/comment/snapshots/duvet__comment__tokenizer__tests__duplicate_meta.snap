---
source: duvet/src/comment/tokenizer.rs
expression: tokens
---
[
    Meta {
        key: "meta",
        value: "1",
        line: 1,
    },
    Meta {
        key: "meta",
        value: "2",
        line: 2,
    },
]
